ckanext/validation/jobs.py

```diff
@@ -54,54 +55,62 @@ def run_validation_job(resource):
         {'ignore_auth': True}, {'id': resource['package_id']})
 
     source = None
-    if resource.get(u'url_type') == u'upload':
+    if resource.get('url_type') == 'upload':
         upload = uploader.get_resource_uploader(resource)
         if isinstance(upload, uploader.ResourceUpload):
-            source = upload.get_path(resource[u'id'])
+            source = upload.get_path(resource['id'])
         else:
             # Upload is not the default implementation (ie it's a cloud storage
             # implementation)
             pass_auth_header = t.asbool(
-                t.config.get(u'ckanext.validation.pass_auth_header', True))
-            if dataset[u'private'] and pass_auth_header:
+                t.config.get('ckanext.validation.pass_auth_header', True))
+            if dataset['private'] and pass_auth_header:
                 s = requests.Session()
                 s.headers.update({
-                    u'Authorization': t.config.get(
-                        u'ckanext.validation.pass_auth_header_value',
+                    'Authorization': t.config.get(
+                        'ckanext.validation.pass_auth_header_value',
                         _get_site_user_api_key())
                 })
 
-                options[u'http_session'] = s
+                options['http_session'] = s
 
     if not source:
-        source = resource[u'url']
-
-    schema = resource.get(u'schema')
-    if schema and isinstance(schema, str):
-        if schema.startswith('http'):
-            r = requests.get(schema)
-            schema = r.json()
+        source = resource['url']
+
+    schema = resource.get('schema')
+    if schema:
+        if isinstance(schema, str):
+            if schema.startswith('http'):
+                r = requests.get(schema)
+                schema = r.json()
         else:
-            schema = json.loads(schema)
+            schema = json.dumps(schema)
```

Comment:
This logic is wrong, `schema` can be:
* A python dict, passed directly to the action (eg tests) 
* A string, in that case it can either be:
   * A URL (we need to get the contents and load the JSON)
   * Something else: we assume is a JSON string, so we need to load it (should be valid because it has passed the validators)
We should check all these cases and end up with a schema dict. I assume `validate()` expects a dict for the schema, so I'm not sure why you are using `dumps()` here

Reply:
It looks to me like this code is handling all of those cases, except that it will not consistently result in a dict.

- Passing in a dict will follow the `else` branch and go through `json.dumps`, producing a string.
- Passing in a URL will follow both `if` branches, producing a JSON object.
- Passing a JSON string will follow the first `if`, but not the second, retaining the original string.